ioredis support because of redis sentinel in high availability scenarios

GoncaloMark opened the pull request:
Currently redis (old node-redis) lib for node doesn't support sentinels, it is planned for version 5 which is still not yet stable. So I created a separate ioredis APL to facilitate using local HA environments that leverage redis sentinels.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/saleor/app-sdk?pullRequest=432) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/saleor/app-sdk?pullRequest=432) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/saleor/app-sdk?pullRequest=432) it.</sub>

lkostrowski commented on package.json:
```diff
@@ -39,6 +39,7 @@
     "@opentelemetry/api": "1.9.0",
     "@opentelemetry/semantic-conventions": "1.30.0",
     "debug": "4.4.0",
+    "ioredis": "^5.6.1",
```
Let's not install this, it's a peer dependency only if you actually use redis. (check peerDependencies)

lkostrowski commented on src/APL/redis/ioredis-apl.ts:
```diff
@@ -0,0 +1,239 @@
+    this.debug("Will get auth data from Redis for %s", saleorApiUrl);
+
+    return this.tracer.startActiveSpan(
+      "RedisAPL.get",
```
Let's ensure string names also share IORedis name

lkostrowski commented:
Have you checked plans for supporting Sentinel in the existing library? Looks like they plan to add it soon
https://github.com/redis/node-redis/issues/302#issuecomment-1930456228

lkostrowski commented on src/APL/redis/index.ts:
Technically, this is a separate APL. Which means you either import RedisAPL or IORedisApl. So it should not share the same import to avoid confusion and allow better tree shaking

lkostrowski reviewed (changes requested):
We also gonna need updated docs 
https://docs.saleor.io/developer/extending/apps/developing-apps/app-sdk/apl#available-apls

And rationale when to use which variant

GoncaloMark commented:
> Have you checked plans for supporting Sentinel in the existing library? Looks like they plan to add it soon [redis/node-redis#302 (comment)](https://github.com/redis/node-redis/issues/302#issuecomment-1930456228)

Yes, I've seen they added it on the v5.next7, but that isn't suitable for production environments that want to leverage HA clusters. Also they usually relase a couple more .rc versions before final release, so this could still be some months away tuntil official release while ioredis supports it for quite some time now, being more mature.

GoncaloMark commented:
> We also gonna need updated docs https://docs.saleor.io/developer/extending/apps/developing-apps/app-sdk/apl#available-apls
> 
> And rationale when to use which variant

Okay, I will be fixing these changes this afternoon, thank you! Will update the docs and rationale behind this idea!

lkostrowski commented:
> > We also gonna need updated docs [docs.saleor.io/developer/extending/apps/developing-apps/app-sdk/apl#available-apls](https://docs.saleor.io/developer/extending/apps/developing-apps/app-sdk/apl#available-apls)
> > And rationale when to use which variant
> 
> Okay, I will be fixing these changes this afternoon, thank you! Will update the docs and rationale behind this idea!
> 
> Edit: @lkostrowski where should i write the rationale for the documentation? I'm writing the integration test and once i get it working I'll commit with the changes requested.

It's enough to put this in docs, so users will understand the difference between these implementations

GoncaloMark commented:
@lkostrowski just commited the requested changes, added the integration tests and added on readme md the steps to use a docker compose for the sentinel integration testing (the main use case for this APL). Hope everything is alright. Should i generate a changesets though?

lkostrowski commented:
Yes, it will be new major change `npx changeset`

lkostrowski commented on .changeset/eight-lights-leave.md:
```diff
@@ -0,0 +1,5 @@
+"@saleor/app-sdk": minor
+---
+
+Added support for ioredis to APL as this enables Sentinel support for High Availability Redis clusters.
```
Please add docs 
https://github.com/saleor/saleor-docs

and link them from the changeset

GoncaloMark replied:
Should I commit a small doc section to that repository to https://github.com/saleor/saleor-docs/blob/main/docs/developer/extending/apps/developing-apps/app-sdk/apl.mdx ? Just to be sure.

lkostrowski commented on package.json:
```diff
@@ -16,7 +16,8 @@
     "lint-staged": "lint-staged",
-    "test:integration": "INTEGRATION=1 vitest run test/integration"
+    "test:integration": "INTEGRATION=1 vitest run test/integration",
+    "test:integration-sentinel": "SENTINEL_INTEGRATION=1 vitest run test/integration"
```
I'm wondering if this is actually good pattern to use envs here. Maybe we should utilize some tags/filtering/vitest projects for that to selectively run them? If we have dedicated Vitest project, we could even start docker containers automatically I believe

https://vitest.dev/guide/filtering

@krzysztofzuraw wdyt?

krzysztofzuraw replied:
Yes - I think we can have either folder with `test/integration` or name of the file `test-name.integration.ts`.

By the way - what about running both integration suites (redis / sentinel) in the same `test:integration`?

lkostrowski replied:
I think once we have projects, we can select which projects to run, and we can run both integration suites with single scriot

krzysztofzuraw replied:
we talked with @lkostrowski and decided that we can move right now with envs in this PR so we wont end up changing how to run tests in PR that adds IORedis 😄

lkostrowski commented:
@NyanKiyoshi can you review? Especially the docker part

lkostrowski commented on test/integration/ioredis/ioredis-apl.test.ts:
```diff
@@ -0,0 +1,95 @@
+const testFn = runIntegrationTests ? describe : describe.skip;
+
+testFn("IORedis APL Integration", () => {
+  const client = new Redis({ port: 6379, host: "localhost" });
```
Should it really hardcode address? Probably should be declared in .env or some other shared config that will be commented enough to understand where it comes from

Also I think it's quite messy with this `INTEGRATION=1`, I think we should create dedicated projects:
- units
- integration-ioredis
- integration-redis

So each is scoped

krzysztofzuraw commented on src/APL/ioredis/ioredis-apl.ts:
```diff
@@ -0,0 +1,239 @@
+
+/**
+ * Redis implementation of the Auth Persistence Layer (APL).
+ * This class provides Redis-based storage for Saleor App authentication data.
```
```suggestion
 * This class provides IORedis-based storage for Saleor App authentication data.
```

Can we also add link to IORedis docs here? If I will be using this in my app it will be awesome to have link in JSDoc 😄

GoncaloMark commented:
Hello, good morning, we’ll be looking into making those changes until noon, thank you for the review :) I’ll add to the PR in a bit.

I can start looking into vitest too projects even though it won’t go in this PR yet and start seeing what I can arrange! I’ll keep you updated.

NyanKiyoshi commented on test/integration/ioredis/docker-compose.yml:
```diff
@@ -0,0 +1,155 @@
+version: "3.8"
+
+services:
+  redis-master:
+    image: redis:latest
+    container_name: redis-master
+    hostname: redis-master
```
"main" and "workers" is the preferred terminology

```suggestion
  redis-main:
    image: redis:latest
    container_name: redis-main
    hostname: redis-main
```

NyanKiyoshi commented on test/integration/ioredis/docker-compose.yml:
```diff
@@ -0,0 +1,155 @@
+        ipv4_address: 172.21.0.3
+
+
+  redis-slave-1:
+    image: redis:latest
+    container_name: redis-slave-1
+    hostname: redis-slave-1
+    depends_on:
+      - redis-master
```
```suggestion
  redis-worker-1:
    image: redis:latest
    container_name: redis-worker-1
    hostname: redis-worker-1
    depends_on:
      - redis-main
```

NyanKiyoshi commented on test/integration/ioredis/docker-compose.yml:
```diff
@@ -0,0 +1,155 @@
+        "--appendonly",
+        "yes",
+        "--replicaof",
+        "redis-master",
```
```suggestion
        "redis-main",
```

NyanKiyoshi commented on test/integration/ioredis/docker-compose.yml:
```diff
@@ -0,0 +1,155 @@
+        ipv4_address: 172.21.0.4
+
+
+  redis-slave-2:
+    image: redis:latest
+    container_name: redis-slave-2
+    hostname: redis-slave-2
+    depends_on:
+      - redis-master
```
```suggestion
  redis-worker-2:
    image: redis:latest
    container_name: redis-worker-2
    hostname: redis-worker-2
    depends_on:
      - redis-main
```

NyanKiyoshi commented on test/integration/ioredis/docker-compose.yml:
```diff
@@ -0,0 +1,155 @@
+        "--appendonly",
+        "yes",
+        "--replicaof",
+        "redis-master",
```
```suggestion
        "redis-main",
```

lkostrowski commented:
Hey @GoncaloMark 

We have had internal discussion about accepting more APLs into the core package. We believe that contributions with more APL implementations are great - that's why APL was designed this way.

However it adds extra burden for the core package itself, while Saleor doesn't really utilize these APLs, still must maintain it.

So we had decided that we'd prefer to keep APLs from now on in the community packages. 

Can you please create your own repository and npm package that will expose this APL? I think it's mainly copy-pasting existing codebase. You can use app-sdk as a peer dependency, to utilize AuthData etc.

We will be happy to mention your package in our docs as community resources

GoncaloMark commented:
> Hey @GoncaloMark
> 
> We have had internal discussion about accepting more APLs into the core package. We believe that contributions with more APL implementations are great - that's why APL was designed this way.
> 
> However it adds extra burden for the core package itself, while Saleor doesn't really utilize these APLs, still must maintain it.
> 
> So we had decided that we'd prefer to keep APLs from now on in the community packages.
> 
> Can you please create your own repository and npm package that will expose this APL? I think it's mainly copy-pasting existing codebase. You can use app-sdk as a peer dependency, to utilize AuthData etc.
> 
> We will be happy to mention your package in our docs as community resources

I understand that stance but I think Saleor out of the box should support High Availability scenarios, I also have other patches for Saleor Core because Redis Clusters with Sentinels aren’t immediately supported for self hosted or even cloud deployment use cases.

lkostrowski commented:
> > Hey @GoncaloMark
> > We have had internal discussion about accepting more APLs into the core package. We believe that contributions with more APL implementations are great - that's why APL was designed this way.
> > However it adds extra burden for the core package itself, while Saleor doesn't really utilize these APLs, still must maintain it.
> > So we had decided that we'd prefer to keep APLs from now on in the community packages.
> > Can you please create your own repository and npm package that will expose this APL? I think it's mainly copy-pasting existing codebase. You can use app-sdk as a peer dependency, to utilize AuthData etc.
> > We will be happy to mention your package in our docs as community resources
> 
> I understand that stance but I think Saleor out of the box should support High Availability scenarios, I also have other patches for Saleor Core because Redis Clusters with Sentinels aren’t immediately supported for self hosted or even cloud deployment use cases.

app-sdk is a library, not the final product. Saleor Apps are actually using DynamoDB at scale and it does provide availability.

Adding RedisIO to app-sdk doesn't really change anything - it will *not* be imported in saleor/apps or any other official Saleor app repository - so to use it, you need to fork app and add such APL on you own. Which brings us to the point where you can actually install it from your own npm package
